Moved HTML entities from DATA section to separate resources file.

rage311 opened the pull request:
### Summary
Move HTML entities from Mojo::Util DATA section to separate resources file (#1409)

### Motivation
These changes were requested.

### References
#1409 

The use of `require` helps to circumvent some circular dependencies that occur otherwise.  I'm not sure if this is exactly what was envisioned, and I'm certainly open to input.

kraih commented:
I really don't like the inline requires. Perhaps the code would be cleaner if we just didn't use `Mojo::File` in this one case?

jberger commented:
> I really don't like the inline requires. Perhaps the code would be cleaner if we just didn't use `Mojo::File` in this one case?

I can agree with that. I'd suggest putting a comment in like "don't use Mojo::File here because of circular dependencies" or somethinf

Grinnz commented:
sibling makes it a little ugly to avoid it, can be cleaned up a bit but: `File::Spec->catfile(File::Basename::dirname(Cwd::realpath(__FILE__)), 'resources', 'html_entities.txt')`

rage311 commented:
I reverted all the `require`s and used @Grinnz solution for a `sibling` equivalent.

kraih commented on lib/Mojo/Util.pm:
```diff
@@ -37,11 +38,23 @@ use constant {
+{
+  # Don't use Mojo::File here due to circular dependencies
+  my $html_path
+    = File::Spec->catfile(File::Basename::dirname(Cwd::realpath(__FILE__)),
```
Why `Cwd::realpath` here?

Grinnz replied:
It handles the edge case where Mojo::Util is loaded via a symlink to still reference the correct location

kraih replied:
How could that ever happen? Until your recent addition of `curfile` we did not use `Cwd::realpath` for any of this and there were never any issues.

rage311 replied:
It does seem to work the same if I remove `Cwd::realpath`.

kraih commented on lib/Mojo/Util.pm:
```diff
@@ -37,11 +38,23 @@ use constant {
+    = File::Spec->catfile(File::Basename::dirname(Cwd::realpath(__FILE__)),
+    'resources', 'html_entities.txt');
+
+  local $/ = undef;
```
Why do this on an extra line instead of using the `do` idiom?

rage311 replied:
Like this?
`my $file_content = do { local $/; <$fh> };`

I wasn't familiar with that pattern, I had to Google it.

kraih replied:
Yes, that's the standard way of using `local $/` actually.

kraih commented on lib/Mojo/Util.pm:
```diff
@@ -7,6 +7,7 @@ use Digest::MD5 qw(md5 md5_hex);
 use Digest::SHA qw(hmac_sha1_hex sha1 sha1_hex);
 use Encode 'find_encoding';
 use Exporter 'import';
+use File::Basename;
```
I'm almost certain there are default exports polluting the namespace now.

karenetheridge replied:
You can use Test::CleanNamespaces in an xt/author test to know for sure.

Grinnz replied:
Yes File::Basename has default exports, can be suppressed with `use File::Basename ()`

rage311 commented:
Imported `File::Basename 'dirname'` only.
Changed to `do { ... }` pattern for setting the record separator and slurping the file.
Removed `Cwd::realpath`.

kraih commented on lib/Mojo/Util.pm:
```diff
@@ -37,11 +38,21 @@ use constant {
+    'resources', 'html_entities.txt');
+
+  open my $html_fh, '<', $html_path
+    or croak 'Unable to open html entities file';
```
I don't like hiding the actual error, especially when it would be such an unexpected exception.

kraih commented on lib/Mojo/Util.pm:
```diff
@@ -37,11 +38,21 @@ use constant {
+  open my $html_fh, '<', $html_path
+    or croak 'Unable to open html entities file';
+  my $entity_lines = do { local $/; <$html_fh> };
+  close $html_fh;
```
Do we gain anything by not just letting it go out of scope here?

rage311 replied:
A bit of explicitness, I suppose, but I don't feel strongly about it either way.

kraih commented:
Thanks, merged.
